### PR TITLE
Binterval recursion fix

### PR DESCRIPTION
**Pull request by patrick-kidger**
Rather than trampolining everything, I've taken the more straightforward approach of inlining a lot of the function indirections that were previously present. This should reduce the size of the call stack that's built up by a factor of something like six.

The code is largely cleaner for this change, as well, which is nice.

It does introduce one additional dependency: [boltons](https://boltons.readthedocs.io/en/latest/), which is a popular stdlib-like library for "things that are missing from the stdlib"; in this case a better LRU cache implementation. Obviously additional dependencies are to be avoided, but I think this one is reasonable.

### CODE REVIEW

File: torchsde/_brownian/brownian_interval.py [hunk 1]
```diff
@@ -106,80 +101,67 @@ def increment_and_levy_area(self):
         return W, H, A
 
     def _increment_and_space_time_levy_area(self):
-        return self._top.increment_and_space_time_levy_area_cache(self)
-
-    def increment_and_space_time_levy_area_uncached(self):
         # TODO: switch this over to a trampoline?
```
Comment (lxuechen) [inline torchsde/_brownian/brownian_interval.py:104]: If the plan is to not go for the trampoline approach, should this TODO be removed?

File: setup.py [hunk 2]
```diff
@@ -67,7 +67,7 @@
                        optional=True)
     ],
     cmdclass={'build_ext': cpp_extension.BuildExtension},
-    install_requires=['torch>=1.6.0', 'blist', 'numpy>=1.19.1'],
+    install_requires=['torch>=1.6.0', 'blist', 'numpy>=1.19.1', 'boltons'],
```
Comment (lxuechen) [inline setup.py:70]: Is there a particular min version for boltons that is required? This is mostly reflecting from our numpy version mismatch discussion a while ago.

File: torchsde/_brownian/brownian_interval.py (hunk 1 above)
Comment (patrick-kidger) [reply torchsde/_brownian/brownian_interval.py:104]: I'm leaving it there for now as I think adding a trampoline may still be useful later on, if any use case gives even deeper call stacks.

File: setup.py (hunk 2 above)
Comment (patrick-kidger) [reply setup.py:70]: Not that I'm aware of / have tried tracking down.

**Pull request merged**